### PR TITLE
Improved RawList API, Implementation and Tests

### PR DESCRIPTION
**Pull request by ilexp**
As noted in #820, the `RawList<T>` class can benefit from newly available ref-returns in C# 7.3 for its indexer implementation. Since this is a breaking change for v4.0, it also was a good opportunity to rework and extend some of its API to better address its core use case.

This PR adds several small improvements, as well as a by-ref indexer for `RawList<T>`. For a detailed list, see the individual commits, which have been structured so each one represents a group of changes.

### CODE REVIEW

File: Source/Core/Duality/Utility/RawList.cs [hunk 1]
```diff
@@ -65,49 +66,40 @@ public int Capacity
...
 
 		/// <summary>
-		/// [GET / SET] A safety-checked index accessor to the lists internal array. Will throw an <see cref="System.IndexOutOfRangeException"/>
+		/// [GET / SET] A safety-checked by-ref index accessor to the lists internal array. Will throw an <see cref="System.IndexOutOfRangeException"/>
 		/// when attempting to access indices exceeding <see cref="Count"/>.
 		/// </summary>
 		/// <param name="index"></param>
-		public T this[int index]
+		/// <returns></returns>
+		public ref T this[int index]
```
Comment (Barsonax) [inline Source/Core/Duality/Utility/RawList.cs:74]: Put a `[DebuggerBrowsable(DebuggerBrowsableState.Never)]` here so its consistent with the explicit interface implementations

**Comment by Barsonax** [comment]
For the performance improvements did you do any testing (benchmarkdotnet?)? Pretty sure this improves the performance but its always good to verify and also know how much it improves it.

File: Source/Core/Duality/Utility/RawList.cs (hunk 1 above)
Comment (ilexp) [reply Source/Core/Duality/Utility/RawList.cs:74]: Now that you mention it, this attribute doesn't make sense for any of the indexers. Not sure why I put them there, removed all of them. Thanks!

**Comment by ilexp** [comment]
Addressed the change request you made, see above.

> For the performance improvements did you do any testing (benchmarkdotnet?)? Pretty sure this improves the performance but its always good to verify and also know how much it improves it.

You're right, but no, I didn't run any benchmarks. Mostly for time reasons and given the very small scope of the two tweaks, I stopped at "pretty sure it's an improvement".

**Comment by Barsonax** [comment]
> Addressed the change request you made, see above.
> 
> > For the performance improvements did you do any testing (benchmarkdotnet?)? Pretty sure this improves the performance but its always good to verify and also know how much it improves it.
> 
> You're right, but no, I didn't run any benchmarks. Mostly for time reasons and given the very small scope of the two tweaks, I stopped at "pretty sure it's an improvement".

Ok well as the changes are pretty small and seem to make sense iam assuming it will be ok :)

File: Source/Core/Duality/Utility/RawList.cs (hunk 1 above)
Comment (Barsonax) [reply Source/Core/Duality/Utility/RawList.cs:74]: Haha, I thought you put them there on purpose :P

**Pull request merged**